contracts/StreamManager.vy

```diff
@@ -17,243 +16,313 @@
 ):
-    self.owner = owner
+    self.controller = controller
     MIN_STREAM_LIFE = min_stream_life
-    self.validators = validators
 
     for token: IERC20 in accepted_tokens:
         self.token_is_accepted[token] = True
 
+    self.validators = validators
+
 
 @external
-def set_validators(validators: DynArray[Validator, MAX_VALIDATORS]):
-    assert msg.sender == self.owner
-    self.validators = validators
+def transfer_control(new_controller: address):
+    # NOTE: can revoke transfer at any time calling this method with `self.controller`
+    assert msg.sender == self.controller
+    self.new_controller = new_controller
+    self.new_controller_proposed = block.timestamp
+
+    log NewControllerProposed(msg.sender, new_controller)
 
 
 @external
-def add_token(token: IERC20):
-    assert msg.sender == self.owner
-    self.token_is_accepted[token] = True
+def accept_control():
+    assert msg.sender == self.new_controller
+    assert block.timestamp - self.new_controller_proposed >= CONTROLLER_ACCEPTANCE_DELAY
+
+    log NewControllerAccepted(self.controller, msg.sender)
+    self.controller = msg.sender
 
 
 @external
-def remove_token(token: IERC20):
-    assert msg.sender == self.owner
-    self.token_is_accepted[token] = False
+def set_capabilities(account: address, capabilities: Ability):
+    if Ability.MODFIY_ACCESS not in self.capabilities[msg.sender]:
+        assert msg.sender == self.controller  # dev: insufficient capability
+
+    self.capabilities[account] = capabilities
 
 
 @external
-def create_stream(
-    token: IERC20,
-    amount_per_second: uint256,
-    reason: Bytes[MAX_REASON_SIZE] = b"",
-    start_time: uint256 = block.timestamp,
-) -> uint256:
-    assert self.token_is_accepted[token]  # dev: token not accepted
-    assert start_time <= block.timestamp  # dev: start time in future
+def set_validators(validators: DynArray[Validator, MAX_VALIDATORS]):
+    if Ability.MODFIY_VALIDATORS not in self.capabilities[msg.sender]:
+        assert msg.sender == self.controller  # dev: insufficient capability
 
-    funded_amount: uint256 = staticcall token.allowance(msg.sender, self)
-    if funded_amount == max_value(uint256):
-        funded_amount = staticcall token.balanceOf(msg.sender)
+    self.validators = validators
 
-    max_stream_life: uint256 = max_value(uint256)
-    for validator: Validator in self.validators:
-        # NOTE: Validator either raises or returns a max stream life to use
-        max_stream_life = min(
-            max_stream_life,
-            extcall validator.validate(msg.sender, token, amount_per_second, reason),
-        )
 
-    assert max_stream_life >= funded_amount // amount_per_second  # dev: max stream life too small
+@external
+def set_token_accepted(token: IERC20, is_accepted: bool):
+    if Ability.MODFIY_TOKENS not in self.capabilities[msg.sender]:
+        assert msg.sender == self.controller  # dev: insufficient capability
+
+    self.token_is_accepted[token] = is_accepted
```

Comment:
thought: There's a potential foot gun here by accepting tokens with different decimals. Presumably that could be possible with an intelligent validator but this could trip up users.

Reply:
yeah, pricing tokens accurately is indeed a concern w.r.t. their decimals, for example when I priced out our silverback stuff, there was consistently a significant figure in the 7th decimal place that could lead to the price being off ~3% when using USDC (6 decimals) vs. USDT.

However, on the flip side you can deploy different validators that use `token` to do their pricing (maybe all USD-stables is 1 validator, all EUR is a 2nd, etc.), and yeah like you said you can just adjust your pricing slightly to ensure that specific decimals don't bite you too hard in the ass. If al vallidators return 0 then the stream creation will revert.

USDC is annoying lol

Reply:
This goes back to your point about better docs, I think that's the resolution here

Reply:
7c25276